test/fees/FeeRevertCases.t.sol

```diff
@@ -0,0 +1,288 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {Test} from 'forge-std/Test.sol';
+import {IERC20} from 'openzeppelin-contracts/contracts/token/ERC20/IERC20.sol';
+import {Router, IRouter} from 'src/Router.sol';
+import {FeeCalculatorBase} from 'src/fees/FeeCalculatorBase.sol';
+import {AaveFlashLoanFeeCalculator} from 'src/fees/AaveFlashLoanFeeCalculator.sol';
+import {NativeFeeCalculator} from 'src/fees/NativeFeeCalculator.sol';
+import {AaveV2FlashLoanCallback, IAaveV2FlashLoanCallback} from 'src/callbacks/AaveV2FlashLoanCallback.sol';
+import {IParam} from 'src/interfaces/IParam.sol';
+import {IAgent} from 'src/interfaces/IAgent.sol';
+import {IAaveV2Provider} from 'src/interfaces/aaveV2/IAaveV2Provider.sol';
+
+contract FeeRevertCasesTest is Test {
+    enum InterestRateMode {
+        NONE,
+        STABLE,
+        VARIABLE
+    }
+
+    address public constant NATIVE = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+    address public constant AAVE_V2_PROVIDER = 0xB53C1a33016B2DC2fF3653530bfF1848a515c8c5;
+    address public constant AAVE_V3_PROVIDER = 0x2f39d218133AFaB8F2B819B1066c7E434Ad94E9e;
+    address public constant USDC = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    address public constant USDT = 0xdAC17F958D2ee523a2206206994597C13D831ec7;
+    address public constant DUMMY_TO_ADDRESS = address(0);
+    bytes4 public constant AAVE_FLASHLOAN_SELECTOR =
+        bytes4(keccak256(bytes('flashLoan(address,address[],uint256[],uint256[],address,bytes,uint16)')));
+    bytes4 public constant NATIVE_FEE_SELECTOR = 0xeeeeeeee;
+    uint256 public constant SKIP = 0x8000000000000000000000000000000000000000000000000000000000000000;
+    uint256 public constant BPS_BASE = 10_000;
+    uint256 public constant SIGNER_REFERRAL = 1;
+    uint256 public constant FEE_RATE = 20;
+
+    address v2Pool = IAaveV2Provider(AAVE_V2_PROVIDER).getLendingPool();
+
+    address public user;
+    address public feeCollector;
+    Router public router;
+    IAgent public userAgent;
+    address public flashLoanFeeCalculator;
+    address public nativeFeeCalculator;
+    IAaveV2FlashLoanCallback public flashLoanCallbackV2;
+
+    // Empty arrays
+    address[] public tokensReturnEmpty;
+    IParam.Fee[] public feesEmpty;
+    IParam.Input[] public inputsEmpty;
+
+    function setUp() external {
+        user = makeAddr('User');
+        feeCollector = makeAddr('FeeCollector');
+        address pauser = makeAddr('Pauser');
+
+        // Depoly contracts
+        router = new Router(makeAddr('WrappedNative'), pauser, feeCollector);
+        vm.prank(user);
+        userAgent = IAgent(router.newAgent());
+        flashLoanFeeCalculator = address(new AaveFlashLoanFeeCalculator(address(router), FEE_RATE, AAVE_V3_PROVIDER));
```

Comment:
想請教 AaveFlashLoanFeeCalculator 用 AAVE_V3_PROVIDER, 下面的 FlashLoanCallbackV2 使用 AAVE_V2_PROVIDER, 是故意有差別還是？

Reply:
FlashLoanCallbackV2 use AAVE_V2_PROVIDER because of it's V2's callback, can't use V3. AaveFlashLoanFeeCalculator can use  AAVE_V2_PROVIDER also but I think V3 is more widely used, so use V3.

Reply:
OK, Maybe here can add a comment otherwise people in the future may be confused.